satpy/dataset/dataid.py

```diff
@@ -732,21 +481,85 @@ def create_filtered_query(dataset_key, filter_query):
     return DataQuery.from_dict(ds_dict)
 
 
-def _update_dict_with_filter_query(ds_dict, filter_query):
+def _update_dict_with_filter_query(ds_dict: dict[str, Any], filter_query: dict[str, Any]) -> None:
     if filter_query is not None:
         for key, value in filter_query.items():
             if value != "*":
                 ds_dict.setdefault(key, value)
 
 
-def _create_id_dict_from_any_key(dataset_key):
-    try:
+def _create_id_dict_from_any_key(dataset_key: DataQuery | DataID | str | numbers.Number) -> dict[str, Any]:
+    if hasattr(dataset_key, "to_dict"):
         ds_dict = dataset_key.to_dict()
-    except AttributeError:
-        if isinstance(dataset_key, str):
-            ds_dict = {"name": dataset_key}
-        elif isinstance(dataset_key, numbers.Number):
-            ds_dict = {"wavelength": dataset_key}
-        else:
-            raise TypeError("Don't know how to interpret a dataset_key of type {}".format(type(dataset_key)))
+    elif isinstance(dataset_key, str):
+        ds_dict = {"name": dataset_key}
+    elif isinstance(dataset_key, numbers.Number):
+        ds_dict = {"wavelength": dataset_key}
+    else:
+        raise TypeError("Don't know how to interpret a dataset_key of type {}".format(type(dataset_key)))
     return ds_dict
+
+
+def update_id_with_query(orig_id: DataID, query: DataQuery) -> DataID:
+    """Update a DataID with additional info from a query used to find it."""
+    query_dict = query.to_dict()
+    if not query_dict:
+        return orig_id
+
+    new_id_dict = orig_id.to_dict()
+    orig_id_keys = orig_id.id_keys
+    for query_key, query_val in query_dict.items():
+        # XXX: What if the query_val is a list?
+        if new_id_dict.get(query_key) is None:
+            new_id_dict[query_key] = query_val
+    # don't replace ID key information if we don't have to
+    id_keys = orig_id_keys if all(key in orig_id_keys for key in new_id_dict) else default_id_keys_config
+    new_id = DataID(id_keys, **new_id_dict)
+    return new_id
+
+
+def _keys_to_compare(sdict: dict, odict: dict, o_is_id: bool, shared_keys: bool) -> set:
```

Comment:
what does `o_is_id` stand for?

Reply:
other is DataID or at least looks like one. I could give the variable a longer name for sure. I went back and forth on how to handle the different cases, but at the end of the day my sanity was preserved by doing simple if/else statements on the DataID-ness of "other". Now that most of the logic is stabilizing I could revisit this. Ideas welcome.